### PR TITLE
Improvements to widenedUpperBound.

### PR DESCRIPTION
**Pull request by smillst**
1. Rename widenUpperBound to widenedUpperBound.
2. Allows checkers more easily change the number of iterations before widening.
3. Specify which parameter to widenedUpperBound is the previous value and which is the new value.
4. Changes to the Value Checker widenedUpperBound to take advantage of the
change from the previous value to the new value.

(The Value Checker changes are included because they test the framework changes.)

### CODE REVIEW

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -379,27 +379,74 @@ public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror
...
+            } else {
+                return lub;
+            }
+        }
+
+        private Range widenRange(Range newRange, Range oldRange, Range lubRange) {
+            if (newRange == null || oldRange == null) {
+                return lubRange;
+            }
+            // If the newRange closer to max than the oldRange, then move the from to the
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: This sentence is rather hard to parse. I think an `is` is missing in the condition. Using mark-up for code would also disambiguate the use of `to` as field name or English word.

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -1516,7 +1563,10 @@ public AnnotationMirror convertIntValToIntRange(AnnotationMirror intValAnno) {
         return createIntRangeAnnotation(Collections.min(intValues), Collections.max(intValues));
     }
 
-    /** Returns a {@code Range} bounded by the values specified in the given annotation. */
+    /**
+     * Returns a {@code Range} bounded by the values specified in the given {@code @Range}
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: Should the `@Range` be `@IntRange`?
What about `@IntRangeFromPositive`?
What annotation is covered by the else branch?
Instead of the else branch, why not make the else code top level? (The earlier two ifs also don't use elses.)

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 3]
```diff
@@ -1539,6 +1589,34 @@ public static Range getRange(AnnotationMirror rangeAnno) {
     }
 
     /**
+     * Returns a {@code Range} bounded by the values specified in the given {@code @IntRange} or
+     * {@code @IntVal} annotation. Returns null if the given annotations is not {@code IntRange} or
+     * {@code IntVal}.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: Why does this method not need to cover `@IntRangeFromPositive`?
Why is null allowed for getRange but not for this method?

File: framework/src/org/checkerframework/framework/type/QualifierHierarchy.java [hunk 4]
```diff
@@ -114,31 +114,41 @@ public abstract boolean isSubtype(
     public abstract AnnotationMirror leastUpperBound(AnnotationMirror a1, AnnotationMirror a2);
 
     /**
-     * Returns whether or not this hierarchy implements {@link #widenUpperBound(AnnotationMirror,
-     * AnnotationMirror)}
+     * Returns the number of iterations dataflow should preform before {@link
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/QualifierHierarchy.java]: preform -> perform

File: framework/src/org/checkerframework/framework/type/QualifierHierarchy.java [hunk 5]
```diff
@@ -114,31 +114,41 @@ public abstract boolean isSubtype(
...
+     * #widenedUpperBound(AnnotationMirror, AnnotationMirror)} is called or -1 if it should never be
+     * called.
      *
-     * @return whether or not this hierarchy implements {@link #widenUpperBound(AnnotationMirror,
-     *     AnnotationMirror)}
+     * <p>Subclasses overriding this method should return some positive number.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/QualifierHierarchy.java]: The specification doesn't state that -1 is the default value. So overriding the method with -1 to turn off widening would seem fine.

File: framework/src/org/checkerframework/framework/type/QualifierHierarchy.java [hunk 6]
```diff
@@ -114,31 +114,41 @@ public abstract boolean isSubtype(
...
+     *
+     * @return the number of iterations dataflow should preform before {@link
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/QualifierHierarchy.java]: preform -> perform

File: framework/src/org/checkerframework/framework/type/QualifierHierarchy.java [hunk 7]
```diff
@@ -114,31 +114,41 @@ public abstract boolean isSubtype(
...
-     * <p>Otherwise, returns the least upper bound of the two annotations.
+     * <p>If the type hierarchy has no infinite ascending chain, returns the least upper bound of
+     * the two annotations.
      *
-     * @param a1 annotation
-     * @param a2 annotation
-     * @return an upper bound that is wider than the least upper bound of a1 and a2 (or the lub if
-     *     the type hierarchy does not require this)
+     * @param newQualifier newest qualifier dataflow computed for some expression
+     * @param previousQualifier the qualifier dataflow computed on the last iteration.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/QualifierHierarchy.java]: No full stop at end

File: framework/src/org/checkerframework/framework/type/QualifierHierarchy.java [hunk 8]
```diff
@@ -114,31 +114,41 @@ public abstract boolean isSubtype(
...
      *
-     * <p>Otherwise, returns the least upper bound of the two annotations.
+     * <p>If the type hierarchy has no infinite ascending chain, returns the least upper bound of
+     * the two annotations.
      *
-     * @param a1 annotation
-     * @param a2 annotation
-     * @return an upper bound that is wider than the least upper bound of a1 and a2 (or the lub if
-     *     the type hierarchy does not require this)
+     * @param newQualifier newest qualifier dataflow computed for some expression
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/QualifierHierarchy.java]: I would use `new` and `previous` instead of `newest` and `last` to be symmetric with the parameter names.

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 9]
```diff
@@ -427,27 +427,76 @@ public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror
...
+                Range lubRange = getRange(lub);
+                Range newRange = getRange(newQualifier);
+                Range oldRange = getRange(previousQualifier);
+                return createArrayLenRangeAnnotation(widenRange(newRange, oldRange, lubRange));
+            } else {
+                return lub;
+            }
+        }
+
+        private Range widenRange(Range newRange, Range oldRange, Range lubRange) {
```
Comment (wmdietlGC) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: For consistency, should this be renamed to widenedRange?

File: dataflow/src/org/checkerframework/dataflow/analysis/Store.java [hunk 10]
```diff
@@ -65,14 +65,16 @@
...
      *   <li>Returns a fresh object which is not aliased yet.
      *   <li>Returns an object of the same (dynamic) type as {@code this}, even if the signature is
      *       more permissive.
      *   <li>Is commutative.
      * </ul>
+     *
+     * @param previous must be the previous store
      */
-    S widenUpperBound(S other);
+    S widenUpperBound(S previous);
```
Comment (wmdietlGC) [inline dataflow/src/org/checkerframework/dataflow/analysis/Store.java]: For consistency, should this be renamed to `widenedUpperBound`?
Add a reference to the other `widenedUpperBound` and explain difference in naming?

File: framework/tests/value/loops/OscillatingLoops.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        int d = 0;
+        do {
+            i++;
+            if (d > 4566) {
+                d = 0;
+            } else {
+                d++;
+            }
+        } while (i < Integer.MAX_VALUE);
+        //        @IntRange(from = 0, to = 4567) int after = d;
```
Comment (wmdietlGC) [inline framework/tests/value/loops/OscillatingLoops.java]: Remove or expand this comment.

File: framework/tests/value/loops/OscillatingLoops.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        int d = 1;
+        while (i < Integer.MAX_VALUE) {
+            i++;
+            if (d > 4566) {
+                d = 0;
+            } else {
+                d++;
+            }
+        }
+        //        @IntRange(from = 0, to = 4567) int after = d;
```
Comment (wmdietlGC) [inline framework/tests/value/loops/OscillatingLoops.java]: Remove or expand this comment.

File: framework/tests/value/loops/WidenedUpperBound.java [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+// to make sure that dataflow reaches a fixed point.
+@SuppressWarnings("value")
+public class WidenedUpperBound {
+
+    void increment() {
+        int forIndex;
+        for (forIndex = 0; forIndex < 4323; forIndex++) {
+            @IntRange(from = 0, to = 4322) int x = forIndex;
+        }
+        ////::error: (assignment.type.incompatible)
```
Comment (wmdietlGC) [inline framework/tests/value/loops/WidenedUpperBound.java]: Remove or expand this comment, throughout this file.
